lib/node.js

```diff
@@ -15,30 +15,138 @@
-const GuardCondition = require('./guard_condition.js');
+
+/**
+ * NodeOptions specify configuration choices during the 
+ * node instantiation process.  
+ * @class
+ */
+class NodeOptions {
+
+  /**
+   * Create a new instance with default property values.
+   * @constructor
+   */
+  constructor() {
+    this._startParameterServices = true;
+    this._parameterOverrides = []; 
+    this._automaticallyDeclareParametersFromOverrides = false;
+  }
+
+  /**
+   * Get the startParameterServices option. 
+   * Default value = true;
+   * @returns {boolean} - 
+   */
+  get startParameterServices() {
+    return this._startParameterServices;
+  }
+
+  /**
+   * Set startParameterServices.
+   * @param {boolean} aBool - True indicates for a node to start it's parameter_service.
+   *    Default value is true.
+   */
+  set startParameterServices(aBool) {
+    this._startParameterServices = aBool;
+  }
+
+  /**
+   * Get the parameterOverrides.
+   * @return {Parameter[]} - An array of Parameters that serve as overrides for a node's default
+   *   parameters. Default = empty array [].
+   */
+  get parameterOverrides() {
+    return this._parameterOverrides;
+  }
+
+  /**
+   * Set the Parameters that will serve to override a node's default parameter settings.
+   * @param {Parameter[]} parameters - array of Parameters. Default is an empty array.
+   */
+  set parameterOverrides(parameters) {
+    this._parameterOverrides = parameters;
+  }
+
+  /**
+   * Get the automaticallyDeclareParametersFromOverrides.
+   * 
+   * @returns {boolean} - True indicates that a node shold declare declare parameters from
+   *    it's parameter-overrides 
+   */
+  get automaticallyDeclareParametersFromOverrides() {
+    return this._automaticallyDeclareParametersFromOverrides;
+  }
+
+  /**
+   * Set automaticallyDeclareParametersFromOverrides. 
+   * When true a node will declare parameters from all 
+   * parameter-overrides.
+   * @param {boolean} aBool - 
+   */
+  set automaticallyDeclareParametersFromOverrides(aBool) {
+    this.automaticallyDeclareParametersFromOverrides = aBool;
+  }
+
+  /**
+   * Return an instance configured with default options.
+   * @returns {NodeOptions} - An instance with default values.
+   */
+  static get defaultOptions() {
+    return new NodeOptions();
+  }
+}
+
+// Parameter event publisher constants
+const PARAMETER_EVENT_MSG_TYPE = 'rcl_interfaces/msg/ParameterEvent';
+const PARAMETER_EVENT_TOPIC = 'parameter_events';
+
 /**
  * @class - Class representing a Node in ROS
  * @hideconstructor
  */
 
 class Node {
-  init(name, namespace) {
+
+  /**
+   * Get the NodeOptions class.
+   * 
+   * @returns {NodeOptions} - 
+   */
+  static get NodeOptions() {
+    return NodeOptions;
+  }
+
+  init(name, namespace, context, options) {
     this._publishers = [];
     this._subscriptions = [];
     this._clients = [];
     this._services = [];
     this._timers = [];
     this._guards = [];
     this._name = name;
```

Comment:
Should we remove `_name` now to avoid accidental usage? I think it's being used in a couple places in this class that should now use `this.name()` instead.

Reply:
yes, I think name should be removed.

Reply:
done